File: src/FullBuild/Model/Anthology.cs
// Copyright (c) 2014, Pierre Chalamet
// All rights reserved.
// 
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are met:
//     * Redistributions of source code must retain the above copyright
//       notice, this list of conditions and the following disclaimer.
//     * Redistributions in binary form must reproduce the above copyright
//       notice, this list of conditions and the following disclaimer in the
//       documentation and/or other materials provided with the distribution.
//     * Neither the name of Pierre Chalamet nor the
//       names of its contributors may be used to endorse or promote products
//       derived from this software without specific prior written permission.
// 
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
// ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
// WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
// DISCLAIMED. IN NO EVENT SHALL PIERRE CHALAMET BE LIABLE FOR ANY
// DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
// (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
// LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
// ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

using System;
using System.Collections.Generic;
using System.Collections.Immutable;
using System.IO;
using System.Linq;
using FullBuild.Helpers;
using Newtonsoft.Json;

namespace FullBuild.Model
{
    internal class Anthology
    {
        private const string AnthologyFileName = "anthology.json";

        [JsonProperty("applications")]
        private readonly IImmutableList<Application> _applications;

        [JsonProperty("binaries")]
        private readonly IImmutableList<Binary> _binaries;

        [JsonProperty("bookmarks")]
        private readonly IImmutableList<Bookmark> _bookmarks;

        [JsonProperty("packages")]
        private readonly IImmutableList<Package> _packages;

        [JsonProperty("projects")]
        private readonly IImmutableList<Project> _projects;

        public Anthology()
            : this(ImmutableList.Create<Project>(), ImmutableList.Create<Binary>(), ImmutableList.Create<Package>(), ImmutableList.Create<Application>(), ImmutableList.Create<Bookmark>())
        {
        }

        private Anthology(IEnumerable<Project> projects, IEnumerable<Binary> binaries, IEnumerable<Package> packages, IEnumerable<Application> applications, IEnumerable<Bookmark> bookmarks)
        {
            _projects = projects.OrderBy(x => x.Guid).ToImmutableList();
            _binaries = binaries.OrderBy(x => x.AssemblyName).ToImmutableList();
            _packages = packages.OrderBy(x => x.Name).ToImmutableList();
            _applications = applications.ToImmutableList();
            _bookmarks = bookmarks.ToImmutableList();
        }

        [JsonIgnore]
        public IEnumerable<Binary> Binaries
        {
            get { return _binaries; }
        }

        [JsonIgnore]
        public IEnumerable<Package> Packages
        {
            get { return _packages; }
        }

        [JsonIgnore]
        public IEnumerable<Project> Projects
        {
            get { return _projects; }
        }

        [JsonIgnore]
        public IEnumerable<Application> Applications
        {
            get { return _applications; }
        }

        [JsonIgnore]
        public IEnumerable<Bookmark> Bookmarks
        {
            get { return _bookmarks; }
        }

        private static IImmutableList<T> AddOrUpdate<T>(T obj, IImmutableList<T> list, Func<T, bool> equals) where T : class
        {
            list = Remove(list, equals).Add(obj);
            return list;
        }

        public Anthology AddOrUpdateBookmark(Bookmark bookmark)
        {
            return new Anthology(_projects,
                                 _binaries,
                                 _packages,
                                 _applications,
                                 AddOrUpdate(bookmark, _bookmarks, x => x.Name == bookmark.Name));
        }

        public Anthology AddOrUpdateProject(Project project)
        {
            return new Anthology(AddOrUpdate(project, _projects, x => x.Guid == project.Guid),
                                 _binaries,
                                 _packages,
                                 _applications,
                                 _bookmarks);
        }

        public Anthology AddOrUpdateBinary(Binary binary)
        {
            return new Anthology(_projects,
                                 AddOrUpdate(binary, _binaries, x => x.AssemblyName.InvariantEquals(binary.AssemblyName)),
                                 _packages,
                                 _applications,
                                 _bookmarks);
        }

        public Anthology AddOrUpdatePackages(Package package)
        {
            var existing = Packages.FirstOrDefault(x => x.Name.InvariantEquals(package.Name));
            var newPackages = _packages;
            if (null != existing)
            {
                var version = package.Version.ParseSemVersion();
                var existingVersion = existing.Version.ParseSemVersion();
                if (existingVersion < version)
                {
                    newPackages = newPackages.Replace(existing, package);
                }
                else
                {
                    return this;
                }
            }
            else
            {
                newPackages = newPackages.Add(package);
            }

            return new Anthology(_projects,
                                 _binaries,
                                 newPackages,
                                 _applications,
                                 _bookmarks);
        }

        private static IImmutableList<T> Remove<T>(IImmutableList<T> list, Func<T, bool> equals) where T : class
        {
            var existing = list.FirstOrDefault(@equals);
            if (null != existing)
            {
                list = list.Remove(existing);
            }

            return list;
        }

        public Anthology RemoveProject(Project project)
        {
            return new Anthology(Remove(_projects, x => x.Guid == project.Guid),
                                 _binaries,
                                 _packages,
                                 _applications,
                                 _bookmarks);
        }

        public Anthology RemoveBinary(Binary binary)
        {
            return new Anthology(_projects,
                                 Remove(_binaries, x => x.AssemblyName.InvariantEquals(binary.AssemblyName)),
                                 _packages,
                                 _applications,
                                 _bookmarks);
        }

        public Anthology RemovePackage(Package package)
        {
            return new Anthology(_projects,
                                 _binaries,
                                 Remove(_packages, x => x.Name.InvariantEquals(package.Name)),
                                 _applications,
                                 _bookmarks);
        }

        public static Anthology Load(DirectoryInfo dir)
        {
            dir.Create();

            var file = dir.GetFile(AnthologyFileName);
            if (! file.Exists)
            {
                return new Anthology();
            }

            var oldJson = File.ReadAllText(file.FullName);
            var anthology = JsonConvert.DeserializeObject<Anthology>(oldJson);
            return anthology;
        }

        public void Save(DirectoryInfo dir)
        {
            dir.Create();

            var file = dir.GetFile(AnthologyFileName);
            var json = JsonConvert.SerializeObject(this, Formatting.Indented);
            File.WriteAllText(file.FullName, json);
        }
    }
}

File: src/FsFullBuild/Anthology.fs
// Copyright (c) 2014-2015, Pierre Chalamet
// All rights reserved.
// 
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are met:
//     * Redistributions of source code must retain the above copyright
//       notice, this list of conditions and the following disclaimer.
//     * Redistributions in binary form must reproduce the above copyright
//       notice, this list of conditions and the following disclaimer in the
//       documentation and/or other materials provided with the distribution.
//     * Neither the name of Pierre Chalamet nor the
//       names of its contributors may be used to endorse or promote products
//       derived from this software without specific prior written permission.
// 
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
// ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
// WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
// DISCLAIMED. IN NO EVENT SHALL PIERRE CHALAMET BE LIABLE FOR ANY
// DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
// (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
// LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
// ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
module Anthology

open System
open System.IO
open Collections
open System.Reflection

type OutputType = 
    | Exe
    | Dll

type AssemblyRef = private AssemblyRef of string
with
    member this.Value = (fun (AssemblyRef x) -> x)this
    static member Bind (name : string) = AssemblyRef (name.ToLowerInvariant())
    static member Bind (assName : AssemblyName) = AssemblyRef.Bind (assName.Name)
    static member Bind (file : FileInfo) =  AssemblyRef.Bind (Path.GetFileNameWithoutExtension(file.Name))

type BookmarkName = BookmarkName of string

type BookmarkVersion = BookmarkVersion of string

type Bookmark = 
    { Name : BookmarkName
      Version : BookmarkVersion }

type PackageVersion = PackageVersion of string
with
    member this.Value = (fun (PackageVersion x) -> x)this

type PackageFramework = PackageFramework of string
with
    member this.Value = (fun (PackageFramework x) -> x)this

type PackageId = private PackageId of string
with
    member this.Value = (fun (PackageId x) -> x)this
    static member Bind (id : string) = PackageId (id.ToLowerInvariant())

type Package = 
    { Id : PackageId
      Version : PackageVersion }

type VcsType = 
    | Git
    | Hg

type RepositoryName = private RepositoryName of string
with
    member this.Value = (fun (RepositoryName x) -> x)this
    static member Bind(name : string) = RepositoryName (name.ToLowerInvariant())

type RepositoryUrl = RepositoryUrl of string
with
    member this.Value = (fun (RepositoryUrl x) -> x)this

type Repository = 
    { Name : RepositoryName
      Vcs : VcsType
      Url : RepositoryUrl }

type ProjectRelativeFile = ProjectRelativeFile of string
with
    member this.Value = (fun (ProjectRelativeFile x) -> x)this

type FrameworkVersion = FrameworkVersion of string
with
    member this.Value = (fun (FrameworkVersion x) -> x)this

type ProjectRef = ProjectRef of Guid
with
    member this.Value = (fun (ProjectRef x) -> x)this
    static member Bind(guid : Guid) = ProjectRef guid

type Project = 
    { Repository : RepositoryName
      RelativeProjectFile : ProjectRelativeFile
      ProjectGuid : ProjectRef
      Output : AssemblyRef
      OutputType : OutputType
      FxTarget : FrameworkVersion
      AssemblyReferences : AssemblyRef set
      PackageReferences : PackageId set
      ProjectReferences : ProjectRef set }

type ApplicationName = ApplicationName of string

type Application = 
    { Name : ApplicationName
      Projects : ProjectRef set }

type Anthology = 
    { Applications : Application set
      Repositories : Repository set
      Bookmarks : Bookmark set 
      Projects : Project set }

    


let (|ToRepository|) (vcsType : string, vcsUrl : string, vcsName : string) = 
    let vcs = match vcsType with
              | "git" -> VcsType.Git
              | "hg" -> VcsType.Hg
              | _ -> failwithf "Unknown vcs type %A" vcsType
    { Vcs = vcs
      Name = RepositoryName vcsName
      Url = RepositoryUrl vcsUrl }
